csi/node_server.go

```diff
@@ -197,8 +204,48 @@ func (ns *NodeServer) NodeUnstageVolume(
 	return nil, status.Error(codes.Unimplemented, "")
 }
 
-func (ns *NodeServer) NodeGetVolumeStats(ctx context.Context, in *csi.NodeGetVolumeStatsRequest) (*csi.NodeGetVolumeStatsResponse, error) {
-	return nil, status.Errorf(codes.Unimplemented, "")
+func (ns *NodeServer) NodeGetVolumeStats(ctx context.Context, req *csi.NodeGetVolumeStatsRequest) (*csi.NodeGetVolumeStatsResponse, error) {
+	existVol, err := ns.apiClient.Volume.ById(req.GetVolumeId())
+	if err != nil {
+		return nil, status.Error(codes.Internal, err.Error())
+	}
+	if existVol == nil {
+		msg := fmt.Sprintf("NodeGetVolumeStats: the volume %q not exists", req.GetVolumeId())
+		logrus.Warn(msg)
+		return nil, status.Error(codes.NotFound, msg)
+	}
+
+	volumePath := req.GetVolumePath()
+	if volumePath == "" {
+		return nil, status.Error(codes.InvalidArgument, "NodeGetVolumeStats volume Volume Path cannot be empty")
+	}
+
```

Comment:
in block mode you can just return the longhorn volume capacity.

```go
	return &csi.NodeGetVolumeStatsResponse{
		Usage: []*csi.VolumeUsage{
			&csi.VolumeUsage{
				Total:     // Longhorn volume capacity
				Unit:      csi.VolumeUsage_BYTES,
			},
		},
	}, nil
```

Reply:
The CSI spec does not mention that the volume is in block mode when `volumePath == ""`. It said:
```
  // It can be any valid path where volume was previously
  // staged or published.
  // It MUST be an absolute path in the root filesystem of
  // the process serving this request.
  // This is a REQUIRED field.
  string volume_path = 2;
```
I am thinking that even when the volume is in block mode, it still has a `valid path where volume was previously staged or published`. How do you think?

Reply:
No idea why it linked to that specific line, I wanted to link to the empty space.

Reply:
@yasker do we keep track of whether a volume is used in block mode?

Reply:
Thank you, I see. 

I am thinking about how to detect when the volume is used as a raw block device in the consuming pod.

My first idea is that we can check to see if the filesystem is `tmpfs` and conclude that the volume is a block device bc it does not have a filesystem? There is a catch with this method: use might use the volume as raw block device in the pod's spec but they created a filesystem on it inside the application which we are unaware of. 

My second idea is we find the pod which use the volume and read the information form the pod's spec. This method seems to be ineffective bc we need to call Kubernetes APIs.

How do you think ?

Reply:
You should be able to do a unix.stat against the path, and if `st_rdev` is set then it's a block device.
https://linux.die.net/man/2/stat
Can you give this a try?

Then just refactor the get stats method to return the volumestats response, either for a block device or a mounted filesystem.

Reply:
Thank you Joshua! I will try it

Reply:
By the way, when testing, I notice that Kubelet does not query the Longhorn CSI plugin for the volumes with `volumeMode: Block`.

Reply:
@joshimoo Thank you for the suggestion! The `unix.stat` func works. I implemented it slightly differently.